chapter_recurrent-modern/machine-translation-and-dataset.ipynb

```diff
@@ -0,0 +1,628 @@
+    "we can still process a minibatch of text sequences\n",
+    "at one time by *truncation* and *padding*.\n",
+    "Suppose that every sequence in the same minibatch\n",
+    "should have the same length `numSteps`.\n",
+    "If a text sequence has fewer than `numSteps` tokens,\n",
+    "we will keep appending the special \"&lt;pad&gt;\" token\n",
+    "to its end until its length reaches `numSteps`.\n",
+    "Otherwise,\n",
+    "we will truncate the text sequence\n",
+    "by only taking its first `numSteps` tokens\n",
+    "and discarding the remaining.\n",
+    "In this way,\n",
+    "every text sequence\n",
+    "will have the same length\n",
+    "to be loaded in minibatches of the same shape.\n",
+    "\n",
+    "The following `truncatePad` function\n",
+    "truncates or pads text sequences as described before.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static int[] truncatePad(Integer[] integerLine, int numSteps, int paddingToken) {\n",
+    "    /* Truncate or pad sequences */\n",
+    "    int[] line = Arrays.stream(integerLine).mapToInt(i -> i).toArray();\n",
+    "    if (line.length > numSteps) return Arrays.copyOfRange(line, 0, numSteps);\n",
+    "    int[] paddingTokenArr = new int[numSteps - line.length]; // Pad\n",
+    "    for (int i = 0; i < paddingTokenArr.length; i++) paddingTokenArr[i] = paddingToken;\n",
+    "\n",
+    "    return IntStream.concat(Arrays.stream(line), Arrays.stream(paddingTokenArr)).toArray();\n",
+    "}\n",
+    "\n",
+    "int[] result = truncatePad(srcVocab.getIdxs(source.get(0)), 10, srcVocab.getIdx(\"<pad>\"));\n",
+    "System.out.println(Arrays.toString(result));"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 16
+   },
+   "source": [
+    "Now we define a function to transform\n",
+    "text sequences into minibatches for training.\n",
+    "We append the special “&lt;eos&gt;” token\n",
+    "to the end of every sequence to indicate the\n",
+    "end of the sequence.\n",
+    "When a model is predicting\n",
+    "by\n",
+    "generating a sequence token after token,\n",
+    "the generation\n",
+    "of the “&lt;eos&gt;” token\n",
+    "can suggest that\n",
+    "the output sequence is complete.\n",
+    "Besides,\n",
+    "we also record the length\n",
+    "of each text sequence excluding the padding tokens.\n",
+    "This information will be needed by\n",
+    "some models that\n",
+    "we will cover later.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static Pair<NDArray, NDArray> buildArrayNMT(\n",
+    "        ArrayList<String[]> lines, Vocab vocab, int numSteps) {\n",
+    "    /* Transform text sequences of machine translation into minibatches. */\n",
+    "    ArrayList<Integer[]> linesIntArr = new ArrayList<>();\n",
```

Comment:
you should be able to achieve the same thing with int[]

Reply:
So I am using the Integer class type in Vocab since I am storing it in a Hashmap. Should I then be converting Integer[] to int[] before returning the values in `getIdxs()`? I try to stick with int[] but whenever I create a List for example, the only way I have seen possible is to use Integer so not sure if what you mean is I should always convert back to int[].

Reply:
Is the HashMap in this notebook? I can't find it.

Reply:
For the HashMap it's in the Vocab class (TimeMachineUtils.java).